sklearn/base.py

```diff
@@ -1087,3 +1088,33 @@ def is_outlier_detector(estimator):
         True if estimator is an outlier detector and False otherwise.
     """
     return getattr(estimator, "_estimator_type", None) == "outlier_detector"
+
+
+def _fit_context(*, skip_nested_validation=True):
```

Comment:
`_fit_context` does not describe what the decorator does. Maybe:

```suggestion
def _validate_param_context(*, prefer_skip_nested_validation):
```

I'm open to better names.

Also, I think `prefer_skip_nested_validation` should be passed in all the time. It makes it easier to understand the indention from the call-site. (See other comment about `prefer_skip_nested_validation`.)

Reply:
I used a generic name on purpose. In the future we might want to run ``fit`` within another context manager, for instance callbacks need that fit is ran in a ``try finally`` to have a proper tear down.

> Also, I think prefer_skip_nested_validation should be passed in all the time.

I'm ok passing it all the time. My original intention was to skip nested by default (and not passing it explicitely) because this is almost always what we want. This way when we pass it explicitely False, it's more visible that this is different. What do you think ?

Reply:
>  My original intention was to skip nested by default (and not passing it explicitely) because this is almost always what we want. This way when we pass it explicitely False, it's more visible that this is different. What do you think ?

Given that the function name is `_fit_context`, I do not think the default of "turning off nested validation" is obviously `True`. For scikit-learn, there is a sizable amount of estimators that still wants nested validation. Off the top of my head:

- All meta-estimators (validating the estimator parameters)
- All estimators that accept a scorer (validating the parameters for the metric)
- All estimators that accept a splitter (validating the splitter parameters)

Reply:
I made it a required param for `_fit_context`.

For functions I left the default to True, otherwise I need to set it for all currently decorated functions as part of this PR. I thought I could do that in a separate PR. Do you prefer making the change for all function here ?